synapse/storage/databases/main/cache.py

```diff
@@ -85,7 +99,13 @@ def get_all_updated_caches_txn(txn):
                 LIMIT ?
             """
             txn.execute(sql, (last_id, instance_name, limit))
-            updates = [(row[0], row[1:]) for row in txn]
+            updates: List[Tuple[int, _CacheData]] = []
+            row: Tuple[int, str, Optional[List[str]], Optional[int]]
+            # Type saftey: iterating over `txn` yields `Tuple`, i.e.
+            # `Tuple[Any, ...]` of arbitrary length. Mypy detects assigning a
+            # variadic tuple to a fixed length tuple and flags it up as an error.
+            for row in txn:  # type: ignore[assignment]
+                updates.append((row[0], row[1:]))
```

Comment:
I think this could probably have stayed a list comprehension?

Reply:
Unfortunately I think the annotation to `row` on 123 doesn't apply to an iteration variable inside a comprehension or generator expression.

Replacing 122--128 with the following

```python
            updates: List[Tuple[int, _CacheData]] = []
            row: Tuple[int, str, Optional[List[str]], Optional[int]]
            updates = [(row[0], row[1:]) for row in txn]
```

yields:

```
mypy
synapse/storage/databases/main/cache.py:127: error: List comprehension has incompatible type List[Tuple[Any, Tuple[Any, ...]]]; expected List[Tuple[int, Tuple[str, Optional[List[str]], Optional[int]]]]  [misc]
```

Reply:
Ah, that's unfortunate. I think it gets scoped to just the comprehension.

Reply:
Yeah. I think we'd need to `cast()` inside the comprehension. I felt like the above was the least-bad option.